File: sdk/base-js/src/crdt/text.ts
/**
 * CRDTText -- collaborative text based on RGA (Replicated Growable Array).
 *
 * Each character has a unique position id. Insertions reference the position
 * id of the character they follow. Deletions tombstone position ids.
 * The structure resolves conflicts by using HLC comparison on concurrent
 * inserts at the same position.
 */

import type { HLCTimestamp } from './clock.js'
import { HybridLogicalClock, compareHLC } from './clock.js'
import type { Operation, TextInsertPayload, TextDeletePayload } from './operations.js'
import { makeOpId, makePositionId } from './operations.js'

// ---------------------------------------------------------------------------
// Internal node structure
// ---------------------------------------------------------------------------

interface TextNode {
  id: string         // unique position id
  char: string       // single character
  hlc: HLCTimestamp  // creation timestamp
  tombstone: boolean // soft-deleted
  afterId: string | null // the id this node was inserted after
}

// ---------------------------------------------------------------------------
// CRDTText
// ---------------------------------------------------------------------------

export type TextChangeCallback = (text: string) => void

export class CRDTText {
  private _nodes: TextNode[] = []
  private readonly _clock: HybridLogicalClock
  private readonly _documentId: string
  private readonly _field: string
  private readonly _pendingOps: Operation[] = []
  private _listeners = new Set<TextChangeCallback>()

  /** Index from position id to array index for O(1) lookup. */
  private _index = new Map<string, number>()

  constructor(documentId: string, field: string, clock: HybridLogicalClock) {
    this._documentId = documentId
    this._field = field
    this._clock = clock
  }

  // ---- Public API ---------------------------------------------------------

  /** Insert text at a visible character position (0-based). */
  insert(position: number, text: string): Operation {
    if (text.length === 0) {
      throw new Error('CRDTText.insert: empty text')
    }

    const afterId = this._visibleIdAtPosition(position - 1)
    const hlc = this._clock.now()
    const positionIds: string[] = []

    // Create nodes for each character, chaining them.
    let prevId = afterId
    for (let i = 0; i < text.length; i++) {
      const posId = makePositionId(hlc, i)
      positionIds.push(posId)

      const node: TextNode = {
        id: posId,
        char: text[i],
        hlc: { ...hlc, counter: hlc.counter + i },
        tombstone: false,
        afterId: prevId,
      }

      this._insertNode(node)
      prevId = posId
    }

    const op: Operation = {
      id: makeOpId(hlc),
      documentId: this._documentId,
      field: this._field,
      hlc,
      type: 'text.insert',
      payload: { afterId, content: text, positionIds } satisfies TextInsertPayload,
    }

    this._pendingOps.push(op)
    this._notifyChange()
    return op
  }

  /** Delete `length` visible characters starting at `position` (0-based). */
  delete(position: number, length: number): Operation {
    const ids = this._visibleIdsInRange(position, length)
    if (ids.length === 0) {
      throw new Error('CRDTText.delete: nothing to delete')
    }

    for (const id of ids) {
      const idx = this._index.get(id)
      if (idx !== undefined) {
        this._nodes[idx].tombstone = true
      }
    }

    const hlc = this._clock.now()
    const op: Operation = {
      id: makeOpId(hlc),
      documentId: this._documentId,
      field: this._field,
      hlc,
      type: 'text.delete',
      payload: { positionIds: ids } satisfies TextDeletePayload,
    }

    this._pendingOps.push(op)
    this._notifyChange()
    return op
  }

  /** Apply a remote operation. */
  applyRemote(op: Operation): void {
    this._clock.receive(op.hlc)

    if (op.type === 'text.insert') {
      const payload = op.payload as TextInsertPayload
      for (let i = 0; i < payload.content.length; i++) {
        const node: TextNode = {
          id: payload.positionIds[i],
          char: payload.content[i],
          hlc: { ...op.hlc, counter: op.hlc.counter + i },
          tombstone: false,
          afterId: i === 0 ? payload.afterId : payload.positionIds[i - 1],
        }
        this._insertNode(node)
      }
    } else if (op.type === 'text.delete') {
      const payload = op.payload as TextDeletePayload
      for (const posId of payload.positionIds) {
        const idx = this._index.get(posId)
        if (idx !== undefined) {
          this._nodes[idx].tombstone = true
        }
      }
    }

    this._notifyChange()
  }

  /** Get the current visible text. */
  toString(): string {
    let result = ''
    for (const node of this._nodes) {
      if (!node.tombstone) {
        result += node.char
      }
    }
    return result
  }

  /** Number of visible characters. */
  get length(): number {
    let count = 0
    for (const node of this._nodes) {
      if (!node.tombstone) count++
    }
    return count
  }

  /** Drain pending local operations. */
  drainOps(): Operation[] {
    return this._pendingOps.splice(0, this._pendingOps.length)
  }

  onChange(callback: TextChangeCallback): () => void {
    this._listeners.add(callback)
    return () => {
      this._listeners.delete(callback)
    }
  }

  // ---- Internal -----------------------------------------------------------

  /**
   * Insert a node into the correct position in the array.
   * RGA rule: among siblings sharing the same afterId, order by HLC descending
   * (newer inserts appear first, pushing older ones right).
   */
  private _insertNode(node: TextNode): void {
    // Find the position of the parent (afterId).
    let parentIdx: number
    if (node.afterId === null) {
      parentIdx = -1 // insert at beginning
    } else {
      const idx = this._index.get(node.afterId)
      if (idx === undefined) {
        // Parent not found -- append at end (out-of-order delivery).
        parentIdx = this._nodes.length - 1
      } else {
        parentIdx = idx
      }
    }

    // Scan right from parentIdx+1 to find insertion point.
    // Skip nodes that were also inserted after the same parent but have a
    // higher HLC (they should appear before this node).
    let insertIdx = parentIdx + 1
    while (insertIdx < this._nodes.length) {
      const existing = this._nodes[insertIdx]
      // Only compare with siblings of the same parent.
      if (existing.afterId !== node.afterId) break
      // Higher HLC goes first (left).
      if (compareHLC(existing.hlc, node.hlc) <= 0) break
      insertIdx++
    }

    // Insert and rebuild index for shifted elements.
    this._nodes.splice(insertIdx, 0, node)
    this._rebuildIndex(insertIdx)
  }

  /** Rebuild the id->index map from `startIdx` onward. */
  private _rebuildIndex(startIdx: number): void {
    for (let i = startIdx; i < this._nodes.length; i++) {
      this._index.set(this._nodes[i].id, i)
    }
  }

  /** Get the position id of the visible character at position `pos`, or null for before-head. */
  private _visibleIdAtPosition(pos: number): string | null {
    if (pos < 0) return null
    let visible = -1
    for (const node of this._nodes) {
      if (!node.tombstone) {
        visible++
        if (visible === pos) return node.id
      }
    }
    // pos is past end -- return last visible id.
    for (let i = this._nodes.length - 1; i >= 0; i--) {
      if (!this._nodes[i].tombstone) return this._nodes[i].id
    }
    return null
  }

  /** Get position ids for `length` visible characters starting at `position`. */
  private _visibleIdsInRange(position: number, length: number): string[] {
    const ids: string[] = []
    let visible = -1
    for (const node of this._nodes) {
      if (node.tombstone) continue
      visible++
      if (visible >= position && visible < position + length) {
        ids.push(node.id)
      }
      if (ids.length === length) break
    }
    return ids
  }

  private _notifyChange(): void {
    if (this._listeners.size === 0) return
    const text = this.toString()
    for (const cb of this._listeners) {
      try {
        cb(text)
      } catch {
        // listener errors must not break notification
      }
    }
  }
}

File: crdt/text.go
package crdt

import (
	"fmt"
	"strings"
	"sync"
)

// OpType identifies an RGA operation kind.
type OpType uint8

const (
	OpInsert OpType = iota
	OpDelete
)

// RGAOp represents a single RGA operation for replication.
type RGAOp struct {
	Type      OpType    `json:"type"`
	ID        RGAID     `json:"id"`
	ParentID  RGAID     `json:"parentId"` // for insert: the ID of the node this is inserted after
	Char      rune      `json:"char"`
	Timestamp Timestamp `json:"timestamp"`
}

// RGAID uniquely identifies a character node in the RGA.
type RGAID struct {
	Seq    uint64 `json:"seq"`
	NodeID NodeID `json:"nodeId"`
}

// IsZero reports whether the ID is the zero/sentinel value.
func (id RGAID) IsZero() bool {
	return id.Seq == 0 && id.NodeID == ""
}

// String returns a human-readable ID.
func (id RGAID) String() string {
	return fmt.Sprintf("%s:%d", id.NodeID, id.Seq)
}

// After reports whether id should be ordered after other (for tie-breaking).
func (id RGAID) After(other RGAID) bool {
	if id.Seq != other.Seq {
		return id.Seq > other.Seq
	}
	return id.NodeID > other.NodeID
}

// rgaNode is an internal linked-list node in the RGA.
type rgaNode struct {
	id      RGAID
	char    rune
	deleted bool
	next    *rgaNode
}

// RGA (Replicated Growable Array) is a CRDT for collaborative text editing.
// It maintains a linked list of character nodes with unique IDs that allow
// concurrent inserts to be ordered deterministically.
type RGA struct {
	mu       sync.RWMutex
	head     *rgaNode          // sentinel head node
	index    map[string]*rgaNode // id.String() -> node for O(1) lookup
	seq      uint64
	nodeID   NodeID
	pending  []RGAOp           // pending ops for sync
}

// NewRGA creates a new RGA instance for the given node.
func NewRGA(nodeID NodeID) *RGA {
	sentinel := &rgaNode{id: RGAID{}} // zero ID sentinel
	return &RGA{
		head:   sentinel,
		index:  map[string]*rgaNode{sentinel.id.String(): sentinel},
		nodeID: nodeID,
	}
}

// nextID generates the next unique ID for this node.
func (r *RGA) nextID() RGAID {
	r.seq++
	return RGAID{Seq: r.seq, NodeID: r.nodeID}
}

// Insert inserts a character after the given position (0-based index).
// Position -1 means insert at the beginning.
// Returns the generated operation for replication.
func (r *RGA) Insert(position int, ch rune) RGAOp {
	r.mu.Lock()
	defer r.mu.Unlock()

	parentNode := r.nodeAtVisible(position)
	id := r.nextID()

	newNode := &rgaNode{id: id, char: ch}
	r.insertAfter(parentNode, newNode)

	op := RGAOp{
		Type:      OpInsert,
		ID:        id,
		ParentID:  parentNode.id,
		Char:      ch,
		Timestamp: Timestamp{Time: int64(id.Seq), NodeID: r.nodeID},
	}
	r.pending = append(r.pending, op)
	return op
}

// Delete deletes the character at the given visible position (0-based).
// Returns the generated operation for replication.
func (r *RGA) Delete(position int) (RGAOp, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	node := r.visibleNodeAt(position)
	if node == nil {
		return RGAOp{}, fmt.Errorf("position %d out of range", position)
	}

	node.deleted = true

	op := RGAOp{
		Type:      OpDelete,
		ID:        node.id,
		Timestamp: Timestamp{Time: int64(r.seq), NodeID: r.nodeID},
	}
	r.pending = append(r.pending, op)
	return op, nil
}

// ApplyOp applies a remote operation. Returns an error if the operation
// references a parent that doesn't exist (operation should be retried later).
func (r *RGA) ApplyOp(op RGAOp) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	switch op.Type {
	case OpInsert:
		// check if already applied
		if _, exists := r.index[op.ID.String()]; exists {
			return nil // idempotent
		}

		parentKey := op.ParentID.String()
		parent, ok := r.index[parentKey]
		if !ok {
			return fmt.Errorf("parent %s not found", parentKey)
		}

		newNode := &rgaNode{id: op.ID, char: op.Char}
		r.insertAfter(parent, newNode)

		// update local seq if remote is ahead
		if op.ID.Seq > r.seq {
			r.seq = op.ID.Seq
		}

	case OpDelete:
		key := op.ID.String()
		node, ok := r.index[key]
		if !ok {
			return fmt.Errorf("node %s not found for delete", key)
		}
		node.deleted = true
	}

	return nil
}

// insertAfter inserts newNode after parent, respecting ordering among siblings.
// Among children of the same parent, nodes with higher IDs come first (left).
func (r *RGA) insertAfter(parent, newNode *rgaNode) {
	// find the correct position: skip nodes that have higher IDs
	// (they were inserted after the same parent but should appear first).
	cursor := parent
	for cursor.next != nil && cursor.next.id.After(newNode.id) {
		cursor = cursor.next
	}

	newNode.next = cursor.next
	cursor.next = newNode
	r.index[newNode.id.String()] = newNode
}

// nodeAtVisible returns the node at the given visible position.
// Position -1 returns the sentinel head.
func (r *RGA) nodeAtVisible(pos int) *rgaNode {
	if pos < 0 {
		return r.head
	}

	node := r.visibleNodeAt(pos)
	if node == nil {
		// past end: find last node
		last := r.head
		for last.next != nil {
			last = last.next
		}
		return last
	}
	return node
}

// visibleNodeAt returns the node at the given 0-based visible position.
func (r *RGA) visibleNodeAt(pos int) *rgaNode {
	idx := -1
	for n := r.head.next; n != nil; n = n.next {
		if !n.deleted {
			idx++
			if idx == pos {
				return n
			}
		}
	}
	return nil
}

// ToString returns the current visible text content.
func (r *RGA) ToString() string {
	r.mu.RLock()
	defer r.mu.RUnlock()

	var sb strings.Builder
	for n := r.head.next; n != nil; n = n.next {
		if !n.deleted {
			sb.WriteRune(n.char)
		}
	}
	return sb.String()
}

// Length returns the number of visible characters.
func (r *RGA) Length() int {
	r.mu.RLock()
	defer r.mu.RUnlock()

	count := 0
	for n := r.head.next; n != nil; n = n.next {
		if !n.deleted {
			count++
		}
	}
	return count
}

// Operations returns pending operations and clears the pending buffer.
func (r *RGA) Operations() []RGAOp {
	r.mu.Lock()
	defer r.mu.Unlock()

	ops := r.pending
	r.pending = nil
	return ops
}

// Merge applies all operations from a remote RGA.
// This is a convenience method that replays operations.
func (r *RGA) Merge(other *RGA) error {
	other.mu.RLock()
	// Collect all nodes from other in order
	var ops []RGAOp
	for n := other.head.next; n != nil; n = n.next {
		op := RGAOp{
			Type: OpInsert,
			ID:   n.id,
			Char: n.char,
		}
		// find parent: the previous node in the linked list
		prev := other.head
		for p := other.head; p.next != nil && p.next != n; p = p.next {
			prev = p.next
		}
		op.ParentID = prev.id
		ops = append(ops, op)

		if n.deleted {
			ops = append(ops, RGAOp{
				Type: OpDelete,
				ID:   n.id,
			})
		}
	}
	other.mu.RUnlock()

	for _, op := range ops {
		if err := r.ApplyOp(op); err != nil {
			return err
		}
	}
	return nil
}

// InsertText inserts a string starting at the given position.
// Returns the operations generated.
func (r *RGA) InsertText(position int, text string) []RGAOp {
	var ops []RGAOp
	for i, ch := range text {
		op := r.Insert(position+i, ch)
		ops = append(ops, op)
	}
	return ops
}

// StateVector returns a map of nodeID -> max sequence seen, used for sync.
func (r *RGA) StateVector() map[NodeID]uint64 {
	r.mu.RLock()
	defer r.mu.RUnlock()

	sv := make(map[NodeID]uint64)
	for n := r.head.next; n != nil; n = n.next {
		if n.id.Seq > sv[n.id.NodeID] {
			sv[n.id.NodeID] = n.id.Seq
		}
	}
	return sv
}

// OpsSince returns all operations for nodes whose sequence is greater than
// the values in the provided state vector.
func (r *RGA) OpsSince(sv map[NodeID]uint64) []RGAOp {
	r.mu.RLock()
	defer r.mu.RUnlock()

	var ops []RGAOp
	for n := r.head.next; n != nil; n = n.next {
		maxSeen := sv[n.id.NodeID]
		if n.id.Seq > maxSeen {
			// find the parent (previous node)
			prev := r.head
			for p := r.head; p.next != nil && p.next != n; p = p.next {
				prev = p.next
			}

			ops = append(ops, RGAOp{
				Type:     OpInsert,
				ID:       n.id,
				ParentID: prev.id,
				Char:     n.char,
			})

			if n.deleted {
				ops = append(ops, RGAOp{
					Type: OpDelete,
					ID:   n.id,
				})
			}
		}
	}
	return ops
}
